### PR TITLE
fix(admin): inject updatetime when the requst is PATCH with sub path

### LINKED ISSUE #4763: bug: using HTTP PATCH method to modify plugin pemeters return 405
### Issue description
on apisx 2.1, it supports using Http PATCH to modify plugin pemeters，for example 
```
[root@qizhendong-dev2 /]# curl  http://127.0.0.1:9080/apisix/admin/global_rules/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -s | python -m json.tool
{
    "action": "get",
    "count": "1",
    "node": {
        "key": "/apisix/global_rules/1",
        "value": {
            "create_time": 1628162770,
            "id": "1",
            "plugins": {
                "http-logger": {
                    "auth_header": "",
                    "batch_max_size": 1,
                    "buffer_duration": 60,
                    "concat_method": "new_line",
                    "inactive_timeout": 5,
                    "include_req_body": false,
                    "max_retry_count": 0,
                    "name": "http logger",
                    "retry_delay": 1,
                    "send_prometheus_data": false,
                    "timeout": 3,
                    "uri": "http://10.181.159.72:15151",
                    "use_tcp_transmission": true
                },
                "limit-conn": {
                    "burst": 0,
                    "conn": 10001,
                    "default_conn_delay": 0.1,
                    "key": "remote_addr",
                    "rejected_code": 503
                }
            },
            "update_time": 1628162770
        }
    }
}
[root@qizhendong-dev2 /]# curl -i -X PATCH  http://127.0.0.1:9080/apisix/admin/global_rules/1/plugins/limit-conn/conn -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -d "10000"
HTTP/1.1 405 Not Allowed
Date: Thu, 05 Aug 2021 11:26:21 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 154
Connection: close
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Access-Control-Expose-Headers: *
Access-Control-Max-Age: 3600
Server: APISIX/2.7

<html>
<head><title>405 Not Allowed</title></head>
<body>
<center><h1>405 Not Allowed</h1></center>
<hr><center>openresty</center>
</body>
</html>
```
error.log
```
2021/08/05 19:26:21 [error] 4705#4705: *1582 lua entry thread aborted: runtime error: /usr/local/apisix/apisix/admin/utils.lua:37: attempt to index local 'patch_conf' (a number value)
stack traceback:
coroutine 0:
        /usr/local/apisix/apisix/admin/utils.lua: in function 'inject_timestamp'
        /usr/local/apisix/apisix/admin/global_rules.lua:159: in function </usr/local/apisix/apisix/admin/global_rules.lua:116>
        /usr/local/apisix/apisix/admin/init.lua:176: in function 'handler'
        /usr/local/apisix//deps/share/lua/5.1/resty/radixtree.lua:721: in function 'dispatch'
        /usr/local/apisix/apisix/init.lua:757: in function 'http_admin'
        content_by_lua(nginx.conf:232):2: in main chunk, client: 127.0.0.1, server: _, request: "PATCH /apisix/admin/global_rules/1/plugins/limit-conn/conn HTTP/1.1", host: "127.0.0.1:9080"

````

**Now, I use 2.1 which can work, but on 2.7 I got this error. We have many deploy scripts to manage our cluter and use a lot of http PATCH method**

### Environment

Bug report without environment information will be ignored or closed.

* apisix version (cmd: `apisix version`): 
/usr/local/openresty/luajit/bin/luajit /usr/local/apisix/apisix/cli/apisix.lua version
2.7
* OS (cmd: `uname -a`):
inux qizhendong-dev2 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
* OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):

nginx version: openresty/1.19.3.2
built by gcc 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC)
built with OpenSSL 1.1.1k  25 Mar 2021
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt='-O2 -DNGX_LUA_ABORT_AT_PANIC -I/usr/local/openresty/zlib/include -I/usr/local/openresty/pcre/include -I/usr/local/openresty/openssl111/include' --add-
* etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
{"id":"47dd2614-0fe2-49ba-956a-f1e1ba707f03","version":"2.7","up_time":708,"boot_time":1628162758,"etcd_version":"3.4.0","hostname":"qizhendong-dev2","last_report_time":1628163418}

etcdctl version: 3.4.13
API version: 3.4

* apisix-dashboard version, if have: no
* luarocks version, if the issue is about installation (cmd: `luarocks --version`):
luarocks 3.7.0
LuaRocks main command-line interface

### Minimal test code / Steps to reproduce the issue

Bug report without steps to reproduce will be ignored or closed.

1.
2.
3.

### PR DESCRIPTION
**Pull request by spacewander**
Fix #4763

Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/admin/upstream4.t [hunk 1]
```diff
@@ -560,3 +560,35 @@ passed
...
+            end
+
+            local code, message = t('/apisix/admin/upstreams/1/retries',
+                 ngx.HTTP_PATCH,
+                 json.encode(1)
+            )
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(message)
```
Comment (zhendongcmss) [inline t/admin/upstream4.t:592]: I suggest add PATCH body is json format like:
```
curl http://127.0.0.1:9080/apisix/admin/upstreams/100/nodes  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -i -X PATCH -d '
{
    "39.97.63.215:80": 1,
    "39.97.63.215:81": 1
}'
```

File: apisix/admin/upstreams.lua [hunk 2]
```diff
@@ -204,12 +204,12 @@ function _M.patch(id, conf, sub_path)
         if code then
             return code, err
         end
+        utils.inject_timestamp(new_value, nil, nil)
```
Comment (zhendongcmss) [inline apisix/admin/upstreams.lua:207]: The laste args is true ?

Comment (spacewander) [reply apisix/admin/upstreams.lua]: Good catch. Fixed in the new commit

File: apisix/admin/utils.lua [hunk 3]
```diff
@@ -33,8 +33,15 @@ local function inject_timestamp(conf, prev_conf, patch_conf)
         end
     end
 
-    -- For PATCH request, the modification is passed as 'patch_conf'
-    if not conf.update_time or (patch_conf and patch_conf.update_time == nil) then
+    if not conf.update_time or
+        -- For PATCH request, the modification is passed as 'patch_conf'
+        -- If the sub path is used, the 'patch_conf' will be a placeholder `true`
```
Comment (zhendongcmss) [inline apisix/admin/utils.lua:38]: I think the place holder is nil

File: apisix/admin/global_rules.lua [hunk 4]
```diff
@@ -150,14 +150,14 @@ function _M.patch(id, conf, sub_path)
         if code then
             return code, err
         end
+        utils.inject_timestamp(node_value, nil, true)
```
Comment (zhendongcmss) [inline apisix/admin/global_rules.lua:153]: The laste args is nil

File: t/admin/upstream4.t (hunk 1 above)
Comment (spacewander) [reply t/admin/upstream4.t:598]: PATCH with json has been covered by https://github.com/apache/apisix/blob/183351cd9f68df706d467f4474ca77eaec398100/t/admin/upstream3.t#L348

File: apisix/admin/global_rules.lua (hunk 4 above)
Comment (zhendongcmss) [reply apisix/admin/global_rules.lua:153]: The seme as below， only upstream is right

File: apisix/admin/utils.lua (hunk 3 above)
Comment (zhendongcmss) [reply apisix/admin/utils.lua:38]: if the path_conf placeholder is true, 
 (patch_conf and (patch_conf == true or patch_conf.update_time == nil))  =>
(     true and (true == true or true.update_time == nil)   ), true.update_time will raise error

File: apisix/admin/global_rules.lua (hunk 4 above)
Comment (spacewander) [reply apisix/admin/global_rules.lua:153]: No. The placeholder is true. See other changed files.

File: apisix/admin/utils.lua (hunk 3 above)
Comment (zhendongcmss) [reply apisix/admin/utils.lua:38]: sorry, if path_conf is true, true.update_time doesn't executed.

**Review by zhendongcmss — APPROVED** [review]

**Comment by zhendongcmss** [comment]
@spacewander  Does it can be merged today ？

File: t/admin/upstream4.t (hunk 1 above)
Comment (zhendongcmss) [reply t/admin/upstream4.t:598]: > PATCH with json has been covered by
> 
> https://github.com/apache/apisix/blob/183351cd9f68df706d467f4474ca77eaec398100/t/admin/upstream3.t#L348

we have auto test cases, why we don't find this issue ?

**Comment by spacewander** [comment]
@zhendongcmss 
It depends on when we have enough approving reviews.

**Review by nic-chen — APPROVED** [review]

File: t/admin/upstream4.t (hunk 1 above)
Comment (tzssangglass) [reply t/admin/upstream4.t:598]: This case is not covered by the test case.

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**